business/authentication/openshift_auth_controller.go

```diff
@@ -0,0 +1,191 @@
+	businessInstantiator func(authInfo *api.AuthInfo) (*business.Layer, error)
+
+	// SessionStore persists the session between HTTP requests.
+	SessionStore SessionPersistor
+}
+
+// NewOpenshiftAuthController initializes a new controller for handling OpenShift authentication, with the
+// given persistor and the given businessInstantiator. The businessInstantiator can be nil and
+// the initialized contoller will use the business.Get function.
+func NewOpenshiftAuthController(persistor SessionPersistor, businessInstantiator func(authInfo *api.AuthInfo) (*business.Layer, error)) *openshiftAuthController {
+	if businessInstantiator == nil {
+		businessInstantiator = business.Get
+	}
+
+	return &openshiftAuthController{
+		businessInstantiator: businessInstantiator,
+		SessionStore:         persistor,
+	}
+}
+
+// Authenticate handles an HTTP request that contains the access_token, expires_in URL parameters. The access_token
+// should be the token that was obtained from the OpenShift OAuth server and expires_in is the expiration date-time
+// of the token. The token is validated by obtaining the information user tied to it. Although RBAC is always assumed
+// when using OpenShift, privileges are not checked here.
+func (o openshiftAuthController) Authenticate(r *http.Request, w http.ResponseWriter) (*UserSessionData, error) {
+	err := r.ParseForm()
+
+	if err != nil {
+		return nil, fmt.Errorf("error parsing form info: %w", err)
+	}
+
+	token := r.Form.Get("access_token")
+	expiresIn := r.Form.Get("expires_in")
+	if token == "" || expiresIn == "" {
+		return nil, errors.New("token is empty or invalid")
+	}
+
+	expiresInNumber, err := strconv.Atoi(expiresIn)
+	if err != nil {
+		return nil, fmt.Errorf("token is empty or invalid: %w", err)
+	}
+
+	expiresOn := time.Now().Add(time.Second * time.Duration(expiresInNumber))
+	bs, err := o.businessInstantiator(&api.AuthInfo{Token: ""})
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving the OAuth package (getting business layer): %w", err)
+	}
+
+	user, err := bs.OpenshiftOAuth.GetUserInfo(token)
+	if err != nil {
+		o.SessionStore.TerminateSession(r, w)
+		return nil, &AuthenticationFailureError{
+			Reason:     "Token is not valid or is expired.",
+			Detail:     err,
+			HttpStatus: http.StatusUnauthorized,
+		}
+	}
+
+	err = o.SessionStore.CreateSession(r, w, config.AuthStrategyOpenshift, expiresOn, openshiftSessionPayload{Token: token})
+	if err != nil {
+		return nil, err
+	}
+
+	return &UserSessionData{
+		ExpiresOn: expiresOn,
+		Username:  user.Metadata.Name,
+		Token:     token,
+	}, nil
+}
+
+// ValidateSession restores a session previously created by the Authenticate function. The user token (access_token)
+// is revalidated by re-fetching user info from the cluster, to ensure that the token hasn't been revoked.
+// If the session is still valid, a populated UserSessionData is returned. Otherwise, nil is returned.
```

Comment:
Yes, we don't have a convention for errors. We give them a mixed usage: both for throwing unexpected errors which we simply throw to the user in logs/ui and also for "expected" recoverable errors which usually are also thrown to the user in a human readable format.

Let's keep this as is, because this auth code is not used across the codebase. It is somewhat specific to only this package.